src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -165,20 +238,52 @@ private async Task PullContainerAsync(IExecutionContext executionContext, Contai
                 string username = string.Empty;
                 string password = string.Empty;
                 string registryType = string.Empty;
+                string authType = string.Empty;
 
                 registryEndpoint.Data?.TryGetValue("registrytype", out registryType);
                 if (string.Equals(registryType, "ACR", StringComparison.OrdinalIgnoreCase))
                 {
+                    try
+                    {
+                        authType = registryEndpoint.Authorization?.Scheme;
+                        Console.WriteLine("Endpoint Authorization Scheme: " + authType);
+                    }
+                    catch { }
+                    try
+                    {
+                        registryEndpoint.Authorization?.Parameters?.TryGetValue("scheme", out authType);
+                        Console.WriteLine("Endpoint Authorization Parameter Scheme: " + authType);
+                    }
+                    catch { }
+                    if (string.IsNullOrEmpty(authType))
+                    {
+                       authType = "ServicePrincipal";
+                    }
+                    authType = "ManagedServiceIdentity";
                     string loginServer = string.Empty;
                     registryEndpoint.Authorization?.Parameters?.TryGetValue("loginServer", out loginServer);
-                    if (loginServer != null) {
+                    if (loginServer != null)
+                    {
                         loginServer = loginServer.ToLower();
                     }
-
-                    registryEndpoint.Authorization?.Parameters?.TryGetValue("serviceprincipalid", out username);
-                    registryEndpoint.Authorization?.Parameters?.TryGetValue("serviceprincipalkey", out password);
-
                     registryServer = $"https://{loginServer}";
+                    if (string.Equals(authType, "ManagedServiceIdentity", StringComparison.OrdinalIgnoreCase))
+                    {
+                        string tenantId = string.Empty;
+                        // Documentation says to pass username through this way
+                        username = "00000000-0000-0000-0000-000000000000";
+                        string AADToken = await GetMSILoginToken(executionContext.CancellationToken);
+                        // change to getting password from string
+                        password = await GetACRPasswordFromAADToken(
+                            AADToken: AADToken, tenantId: tenantId, registryServer: registryServer,
+                            retryCount: 0, timeToWait: 0, cancellationToken: executionContext.CancellationToken);
```

Comment:
When would someone pass custom values to retryCount and timeToWait ? Can they be made as default parameters given this is private function